### PR TITLE
osd/OSD: cleanup OSDService::maybe_share_map()

### PR DESCRIPTION
**Pull request by Matan-B**
```
* avoid locking sent_epoch_lock twice

* add log message when osdmap is not newer than session's projected_epoch

* edit log messages

* rename last_sent_epoch to projected_epoch
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -7284,33 +7284,37 @@ void OSDService::maybe_share_map(
...
-  epoch_t last_sent_epoch = session->last_sent_epoch;
-  session->sent_epoch_lock.unlock();
-
-  if (osdmap->get_epoch() <= last_sent_epoch) {
+  session->projected_epoch_lock.lock();
+  if (peer_epoch_lb > session->projected_epoch) {
+    dout(10) << fmt::format(
+       "{}: con {} updating session's projected_epoch"
+       " of {} from to ping map epoch of {}",
+       __func__, con->get_peer_addr(),
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7292]: "from to" ??

Comment (ronen-fr) [reply src/osd/OSD.cc:7292]: apart from that: LGTM

Comment (Matan-B) [reply src/osd/OSD.cc:7292]: Thanks!

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -7292,25 +7292,20 @@ void OSDService::maybe_share_map(
 	     << " -> " << peer_epoch_lb << " (as per caller)" << dendl;
     session->last_sent_epoch = peer_epoch_lb;
   }
-  epoch_t last_sent_epoch = session->last_sent_epoch;
-  session->sent_epoch_lock.unlock();
 
-  if (osdmap->get_epoch() <= last_sent_epoch) {
+  if (osdmap->get_epoch() <= session->last_sent_epoch) {
```
Comment (rzarzynski) [inline src/osd/OSD.cc]: For the sake of explicitness: I'm assuming no locking in `get_epoch()`.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -7292,25 +7292,20 @@ void OSDService::maybe_share_map(
...
+    session->sent_epoch_lock.unlock();
     return;
   }
 
-  send_incremental_map(last_sent_epoch, con, osdmap);
-  last_sent_epoch = osdmap->get_epoch();
-
-  session->sent_epoch_lock.lock();
-  if (session->last_sent_epoch < last_sent_epoch) {
+  const epoch_t send_from = session->last_sent_epoch;
```
Comment (rzarzynski) [inline src/osd/OSD.cc]: :+1:

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -7292,25 +7292,20 @@ void OSDService::maybe_share_map(
 	     << " -> " << peer_epoch_lb << " (as per caller)" << dendl;
     session->last_sent_epoch = peer_epoch_lb;
   }
-  epoch_t last_sent_epoch = session->last_sent_epoch;
-  session->sent_epoch_lock.unlock();
 
-  if (osdmap->get_epoch() <= last_sent_epoch) {
+  if (osdmap->get_epoch() <= session->last_sent_epoch) {
+    session->sent_epoch_lock.unlock();
```
Comment (rzarzynski) [inline src/osd/OSD.cc]: Would love to handle the unlocking with a scope guard.

**Comment by athanatos** [comment]
Since we're touching this code, let's use a scope guard rather than explicitly locking and unlocking.

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -7292,25 +7292,20 @@ void OSDService::maybe_share_map(
 	     << " -> " << peer_epoch_lb << " (as per caller)" << dendl;
     session->last_sent_epoch = peer_epoch_lb;
   }
-  epoch_t last_sent_epoch = session->last_sent_epoch;
```
Comment (athanatos) [inline src/osd/OSD.cc:7295]: The commit message in this commit needs to be explicit about why this is allowed.  Invoking send_incremental_map doesn't wait until the replica actually receives the map, so marking last_epoch_sent prior to invoking it is just as correct.

**Review by athanatos — CHANGES_REQUESTED** [review]
Looks reasonable to me otherwise.

File: src/osd/OSD.cc [hunk 6]
```diff
@@ -7284,36 +7284,38 @@ void OSDService::maybe_share_map(
...
-  if (peer_epoch_lb > session->projected_epoch) {
-    dout(10) << fmt::format(
-       "{}: con {} updating session's projected_epoch"
-       " from {} to ping map epoch of {}",
-       __func__, con->get_peer_addr(),
-       session->projected_epoch, peer_epoch_lb)
-       << dendl;
-    session->projected_epoch = peer_epoch_lb;
-  }
+  epoch_t send_from;
```
Comment (athanatos) [inline src/osd/OSD.cc:7287]: Initialize this to something.  If there's a constand for an invalid epoch use that.  Otherwise, 0.

**Review by athanatos — APPROVED** [review]
LGTM otherwise

**Comment by athanatos** [comment]
Add needs-qa once the nit above is addressed.

**Comment by ljflores** [comment]
Marking this as TESTED for now, but go ahead and re-add the "needs-qa" label as needed.

**Comment by ljflores** [comment]
Rados suite review: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcCcKXkHLe1789-wip-yuri3-testing-2023-06-19-1518

**Pull request merged**